json-rpc/src/client.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::views::AccountView;
+use anyhow::{ensure, format_err, Error, Result};
+use libra_types::{account_address::AccountAddress, transaction::SignedTransaction};
+use reqwest::Client;
+use serde_json::Value;
+use std::convert::TryFrom;
+
+#[derive(Default)]
+pub struct JsonRpcBatch {
+    pub requests: Vec<(String, Vec<Value>)>,
+}
+
+impl JsonRpcBatch {
+    pub fn add_request(&mut self, method_name: String, parameters: Vec<Value>) {
+        self.requests.push((method_name, parameters));
+    }
+
+    pub fn add_submit_request(&mut self, transaction: SignedTransaction) -> Result<()> {
+        let txn_payload = hex::encode(lcs::to_bytes(&transaction)?);
+        self.add_request("submit".to_string(), vec![Value::String(txn_payload)]);
+        Ok(())
+    }
+
+    pub fn add_get_account_state_request(&mut self, address: AccountAddress) {
+        self.add_request(
+            "get_account_state".to_string(),
+            vec![Value::String(address.to_string())],
+        );
+    }
+}
+
+pub struct JsonRpcAsyncClient {
+    address: String,
+    client: Client,
+}
+
+impl JsonRpcAsyncClient {
+    pub fn new(host: &str, port: u16) -> Self {
+        let address = format!("http://{}:{}", host, port);
+        let client = Client::new();
+        Self { address, client }
+    }
+
+    pub async fn execute(&self, batch: JsonRpcBatch) -> Result<Vec<Result<JsonRpcResponse>>> {
+        let requests = batch
+            .requests
+            .iter()
+            .enumerate()
+            .map(|(id, (method, params))| {
+                serde_json::json!({"jsonrpc": "2.0", "method": method, "params": params, "id": id})
+            })
+            .collect();
+
+        let resp = self
+            .client
+            .post(&self.address)
+            .json(&serde_json::Value::Array(requests))
+            .send()
+            .await?;
+
+        ensure!(
+            resp.status() == 200,
+            format!("Http error code {}", resp.status())
+        );
+
+        let responses: Vec<Value> = resp.json().await?;
+        let mut result = vec![];
+        for _ in batch.requests.iter() {
+            result.push(Err(format_err!("response is missing")));
+        }
+
+        for response in responses {
+            if let Ok(req_id) = self.fetch_id(&response) {
+                if req_id < result.len() {
```

Comment:
Do you want to return error or do smth on `else` for sanity check?

Reply:
vector of results is initialized with errors. error is default if not overwritten by success
also if that ever happens, it means that protocol handler broke. Should be covered by unittests that are testing protocol